NicheTensor API V2

larry-neil-1206 opened the pull request:
NicheTensor API V2 Functions
- Add rate limits unique to each API key, so no one API key can spam unlimited
and exhaust the network.
- Add credits to API keys, so they can only send a limited number of requests
before running out of credits and needing to buy more.
- Different credit costs for different models. For example, the GoJourney model
should have a higher credit cost to use than RealitiesEdgeXL, since it is more
expensive to run.

larry-neil-1206 committed:
feat: Update README.md

larry-neil-1206 committed:
feat: add api-test.py for testing APIs

larry-neil-1206 committed:
feat: integrate credit function and update txt2img function to handle all kinds of models

larry-neil-1206 committed:
feat: complete rate limit function

larry-neil-1206 committed:
feat: merged with master and resolved some conflicts

larry-neil-1206 committed:
feat: minor changes on README.md and requirements.txt files

larry-neil-1206 committed:
feat: specified the python package versions on requirements.txt file

larry-neil-1206 committed:
feat: update README.md

toilaluan commented on api-test.py:
```diff
@@ -0,0 +1,144 @@
+endpoint = "http://localhost:8000/api/v1"
+
+headers = {
+    "API_KEY": "poih34p89bewrfgf3",
```
is it just a dummy api-key?

larry-neil-1206 replied:
It's just a dummy api-key. We can use it as an environment variable or we can get the api-key from user input.

FYI, this api-test.py is a test file to test the API functions on the local environment.
We can add api-test.py to the gitignore list.

toilaluan commented on app.py:
```diff
@@ -76,25 +82,26 @@ class ValidatorInfo(BaseModel):
     def __init__(self):
         self.subtensor = bt.subtensor("finney")
         self.metagraph = self.subtensor.metagraph(23)
+        # mongoDBConnectUri = f"mongodb://{MONGO_DB_USERNAME}:{MONGO_DB_PASSWORD}@localhost:27017"
```
How about the case mongodb is server by other VM, Atlas?

larry-neil-1206 replied:
Yeah, in that case, we need to use the commented Uri as the mongoDBConnectUri. I commented it to test on my local environment.

toilaluan commented on app.py:
```diff
@@ -104,23 +111,24 @@ def __init__(self):
         self.public_key_bytes = self.public_key.public_bytes(
             encoding=serialization.Encoding.Raw, format=serialization.PublicFormat.Raw
         )
+        self.model_list = self.model_config.find_one({"name": "model_list"})["data"]
```
Should we need to add a background process to update these kind of mongodb config for every 10 minutes example?

larry-neil-1206 replied:
That's a good idea.

toilaluan commented on app.py:
```diff
@@ -104,23 +111,24 @@ def __init__(self):
-        self.app.add_api_route(
-            "/api/v1/instantid", self.instantid_api, methods=["POST"]
-        )
+        # self.app.add_api_route(
```
Should you clean the code by removing these commented codes?

larry-neil-1206 replied:
Sure, I will.

toilaluan commented on app.py:
```diff
@@ -243,10 +256,11 @@ async def generate(self, prompt: Union[Prompt, TextPrompt]):
-            for hotkey, log in self.available_validators.items()
-            if log["is_active"]
+            for hotkey, log in validatorItems
+            # if log["is_active"]
```
why we don't check is_active?

larry-neil-1206 replied:
I commented it to test on my local environment. Will uncomment it back.

larry-neil-1206 committed:
Updated some codes according to the feedback
feat: remove dummy api-key from the api-test.py
feat: recover mongoDBConnectUri by uncommencting the original one and commencting the local one
feat: removed useless commented code
feat: uncommented code for checking is_active from validator

larry-neil-1206 committed:
feat: update model_config every 10 mins within thread

larry-neil-1206 committed:
feat: merge with the master branch, resolved some conflicts, and add API_RATE_LIMIT constant

vietbeu commented on app.py:
```diff
@@ -78,25 +83,26 @@ class ValidatorInfo(BaseModel):
-            self.client["image_generation_service"].create_collection("private_key")
+            self.client["image_generation_service"].create_collection(CollectionName.VALIDATORS.value)
+            self.client["image_generation_service"].create_collection(CollectionName.AUTH_KEYS.value)
+            self.client["image_generation_service"].create_collection(CollectionName.PRIVATE_KEY.value)
```
You should create MODEL_CONFIG collection if not exists

vietbeu commented on app.py:
```diff
@@ -584,3 +591,39 @@ def resize_divisible(self, image, max_size=1024, divisible=16):
+
+@app.app.post("/api/v1/controlnet")
+@limiter.limit(API_RATE_LIMIT) # Update the rate limit
+async def controlnet_api(request: Request, data: TextToImage):
```
Data model is incorrect. For example, with /img2img, /controlnet it should be ImgToImg.

larry-neil-1206 replied:
Good catch!

vietbeu commented on app.py:
```diff
@@ -106,30 +112,25 @@ def __init__(self):
         self.public_key_bytes = self.public_key.public_bytes(
             encoding=serialization.Encoding.Raw, format=serialization.PublicFormat.Raw
         )
+        self.model_list = self.model_config.find_one({"name": "model_list"})["data"]
```
It should be 
self.model_list = self.model_config.find_one({"name": "model_list"})
self.model_list = self.model_list.get("data", []) if self.model_list else [] 
In case self.model_config.find_one({"name": "model_list"}) return None

larry-neil-1206 committed:
feat: update APIs route's model interfaces
feat: create model_config collection if it doesn't exist

larry-neil-1206 committed:
feat: handled model_config collection empty exception

larry-neil-1206 committed:
feat: update gitignore file

vietbeu commented on app.py:
```diff
@@ -106,30 +113,25 @@ def __init__(self):
         self.public_key_bytes = self.public_key.public_bytes(
             encoding=serialization.Encoding.Raw, format=serialization.PublicFormat.Raw
         )
+        self.model_list = self.model_config.find_one({"name": "model_list"})["data"] if self.model_config.find_one({"name": "model_list"}) else {}
```
should avoid query DB 2 times

larry-neil-1206 replied:
Yeah, you are right. I missed it. 👍

larry-neil-1206 committed:
feat: update code to avoid DB query 2 times

toilaluan committed:
fix /generate input args

toilaluan committed:
Update constants.py

toilaluan committed:
fill credit for previous keys

toilaluan committed:
hotfix v1

toilaluan committed:
Update constants.py

toilaluan committed:
Merge pull request #4 from SocialTensor/will-dev

update v2: production checked
